File: Python/13179_Smarter_than_hunger.py
from sys import stdin

B = [None]*102

N, R, C,F = None,None,None,None

def cprint(m1,m2,n,rt1,rt2,a):
    node = str(n)
    if n == 0: node = "I"
    if n == F: node = "F"
    print("n:",node,"R1:",m1, "R2:",m2, "rt1:",rt1,"rt2:",rt2,"cans:",a)

def dist(inda,indb):
    return abs((B[inda][0]-B[indb][0]))+abs((B[inda][1]-B[indb][1]))

memo = dict()

def phi(m1,m2,rt1,rt2):
    global memo
    if (m1,m2,rt1,rt2) in memo:
        ans = memo[(m1,m2,rt1,rt2)]
    else:
        if m1 == F and m2 == F:
            ans = 1
        else:
            if (m1 == F):
                cost = max(dist(m2, F)-rt2, 0)
                ans = cost + phi(m1,F,0,0)
            elif (m2 == F):
                cost = max(dist(m1, F)-rt1, 0)
                ans = cost + phi(F,m2,0,0)
            else:
                n = max(m1,m2)
                dm1=max(dist(m1, n+1)-rt1,0)
                dm2=max(dist(m2, n+1)-rt2,0)
                ans = min(dm1 + phi(n+1,m2,0,dm1+rt2), dm2+phi(m1,n+1,dm2+rt1,0))
        memo[(m1,m2,rt1,rt2)] = ans
    return ans

def main():
    global N, R, C, F, B, memo
    data = stdin.readline().strip()
    while len(data) > 0:
        R,C = map(int, data.split())
        N = int(stdin.readline())
        B[0] = (1,1)
        memo = dict()
        for n in range(1,N+1):
            a,b = map(int, stdin.readline().split())
            B[n] = (a,b)
        F = N+1
        B[F] = (R,C)
        print(phi(0,0,0,0))
        data = stdin.readline().strip()    
    
main()

"""
2 4
4
1 1
2 4
1 4
2 2
"""

File: Cpp/13179_Smarter_than_hunger.cpp
#include <iostream>
#include <string>

using namespace std;

pair<int, int> B[102];
int F;

int dist(int inda, int indb) {
    return abs(B[inda].first - B[indb].first) + abs(B[inda].second - B[indb].second);
}

int dp[103][103][151];

int phitab(int RT)
{
    for (int m1 = F; m1>= 0; m1--)
    {
        for (int m2 = F; m2>= 0; m2--)
        {
            for (int rt = RT; rt >= 0; rt--)
            {
                if (m1 == F && m2 == F) {
                    dp[m1][m2][rt] = 1;
                } else {
                    if (m1 == F) {
                        int cost = max(dist(m2, F) - rt, 0);
                        dp[m1][m2][rt] = cost + dp[m1][F][0];
                    } else if (m2 == F) {
                        int cost = max(dist(m1, F) - rt, 0);
                        dp[m1][m2][rt] = cost + dp[F][m2][0];
                    } else {
                        int n = max(m1, m2);
                        int rt1 = 0, rt2 = 0;
                        if (m1<m2)
                        {   
                            rt1 = rt;
                        }else{
                            rt2 = rt;
                        }
                        int dm1 = max(dist(m1, n + 1) - rt1, 0);
                        int dm2 = max(dist(m2, n + 1) - rt2, 0);
                        dp[m1][m2][rt] = min(dm1 + dp[n+1][m2][dm1+rt2], dm2 + dp[m1][n+1][dm2+rt1]);
                    }
                }
            }
        }
    }
    return dp[0][0][0];
}

int main() {
    string line;
    int R, C, N;
    while (getline(cin, line) && !line.empty()) {
        sscanf(line.c_str(), "%d %d", &R, &C);
        cin >> N;
        B[0] = {1, 1};

        for (int n = 1; n <= N; ++n) {
            int a, b;
            cin >> a >> b;
            B[n] = {a, b};
        }

        F = N + 1;
        B[F] = {R, C};

        cout << phitab(min(R*C, 150)) << endl;

        cin.ignore();
    }

    return 0;
}
